p2p/host/resource-manager/limit_defaults.go

```diff
@@ -108,53 +110,295 @@ func (cfg *ScalingLimitConfig) AddProtocolPeerLimit(proto protocol.ID, base Base
-	ServicePeer        map[string]BaseLimit `json:",omitempty"`
+	ServicePeerDefault *ResourceLimits           `json:",omitempty"`
+	ServicePeer        map[string]ResourceLimits `json:",omitempty"`
 
-	ProtocolDefault BaseLimit                 `json:",omitempty"`
-	Protocol        map[protocol.ID]BaseLimit `json:",omitempty"`
+	ProtocolDefault *ResourceLimits                `json:",omitempty"`
+	Protocol        map[protocol.ID]ResourceLimits `json:",omitempty"`
 
-	ProtocolPeerDefault BaseLimit                 `json:",omitempty"`
-	ProtocolPeer        map[protocol.ID]BaseLimit `json:",omitempty"`
+	ProtocolPeerDefault *ResourceLimits                `json:",omitempty"`
+	ProtocolPeer        map[protocol.ID]ResourceLimits `json:",omitempty"`
 
-	PeerDefault BaseLimit             `json:",omitempty"`
-	Peer        map[peer.ID]BaseLimit `json:",omitempty"`
+	PeerDefault *ResourceLimits            `json:",omitempty"`
+	Peer        map[peer.ID]ResourceLimits `json:",omitempty"`
 
-	Conn   BaseLimit `json:",omitempty"`
-	Stream BaseLimit `json:",omitempty"`
+	Conn   *ResourceLimits `json:",omitempty"`
+	Stream *ResourceLimits `json:",omitempty"`
 }
 
-func (cfg *LimitConfig) MarshalJSON() ([]byte, error) {
+func resourceLimitsMapFromBaseLimitMapWithDefaults[K comparable](m map[K]BaseLimit, defaultLimits map[K]BaseLimit, fallbackDefault BaseLimit) map[K]ResourceLimits {
+	if len(m) == 0 {
+		return nil
+	}
+
+	out := make(map[K]ResourceLimits, len(m))
+	for k, v := range m {
+		def := fallbackDefault
+		if defaultForKey, ok := defaultLimits[k]; ok {
+			def = defaultForKey
+		}
+		rl := v.ToResourceLimitsWithDefault(def)
+		if rl != nil {
+			out[k] = *rl
+		}
+	}
+	return out
+}
+
+func (cfg *PartialLimitConfig) MarshalJSON() ([]byte, error) {
 	// we want to marshal the encoded peer id
-	encodedPeerMap := make(map[string]BaseLimit, len(cfg.Peer))
+	encodedPeerMap := make(map[string]ResourceLimits, len(cfg.Peer))
 	for p, v := range cfg.Peer {
 		encodedPeerMap[p.String()] = v
 	}
 
-	type Alias LimitConfig
+	type Alias PartialLimitConfig
 	return json.Marshal(&struct {
 		*Alias
-		Peer map[string]BaseLimit `json:",omitempty"`
+		Peer map[string]ResourceLimits `json:",omitempty"`
 	}{
 		Alias: (*Alias)(cfg),
 		Peer:  encodedPeerMap,
 	})
 }
 
-func (cfg *LimitConfig) Apply(c LimitConfig) {
+func applyResourceLimitsMap[K comparable](this *map[K]ResourceLimits, other map[K]ResourceLimits, fallbackDefault *ResourceLimits) {
+	for k, l := range *this {
+		r := fallbackDefault
+		if l2, ok := other[k]; ok {
+			r = &l2
+		}
+		l.Apply(r)
+		(*this)[k] = l
+	}
+	if *this == nil && other != nil {
+		*this = make(map[K]ResourceLimits)
+	}
+	for k, l := range other {
+		if _, ok := (*this)[k]; !ok {
+			(*this)[k] = l
+		}
+	}
+}
+
+func (cfg *PartialLimitConfig) Apply(c PartialLimitConfig) {
 	cfg.System.Apply(c.System)
 	cfg.Transient.Apply(c.Transient)
 	cfg.AllowlistedSystem.Apply(c.AllowlistedSystem)
```

Comment:
These lines can cause nil pointer exceptions now.

Reply:
zero values for `ResourceLimits` is `DefaultLimit`.

Reply:
I don't understand your comment. Am I missing something?

Reply:
No, I was missing something. I thought that zero value will be zero, but is -2 (BlockAllLimit)